manifest-referrers-api.md

```diff
@@ -0,0 +1,109 @@
+
+```rest
+GET /v2/_ext/oci-artifacts/v1-rc1/{repository}/manifests/{digest}/referrers?n=10
+```
+
+**expanded example:**
+
+```rest
+GET /v2/_ext/oci-artifacts/v1-rc1/net-monitor/manifests/sha256:3c3a4604a545cdc127456d94e421cd355bca5b528f4a9c1905b15da2eb4a4c6b/referrers?n=10
+```
+
+The `/referrers` API MAY provide for filtering of referenceTypes. Artifact clients MUST account for [distribution-spec][oci-distribution-spec] implementations that MAY NOT support filtering. Artifact clients MUST revert to client side filtering to determine which referenceTypes they will process.
+
+### Request Artifacts of a specific media type
+
+**template:**
+```rest
+GET /v2/_ext/oci-artifacts/v1-rc1/{repository}/manifests/{digest}/referrers?n=10&referenceType={referenceType}
+```
+
+**expanded example:**
+
+```rest
+GET /v2/_ext/oci-artifacts/v1-rc1/net-monitor/manifests/sha256:3c3a4604a545cdc127456d94e421cd355bca5b528f4a9c1905b15da2eb4a4c6b/references?n=10&referenceType=application/vnd.oci.notary.v2
+```
+
+### Artifact Referrers API results
+
+[distribution-spec][oci-distribution-spec] implementations MAY implement `referenceType` filtering. Some artifacts types including Notary v2 signatures, may return multiple signatures of the same `referenceType`. To avoid an artifact client from having to retrieve each manifest, just to determine if it's the specific artifact needed to continue processing, the `/referrers` API will return a collection of manifests, including the annotations within each manifest. By providing manifests, as opposed to manifest descriptors, a specific artifact client can find the relevant properties they need to determine which artifact to retrieve. For example, Notary v2 MAY use an annotation: `"org.cncf.notary.v2.signature.subject": "wabbit-networks.io"`, which the client could use to determine which signature to pull from the registry. Using annotations, clients can reduce round trips and the data returned to determine which artifacts the specific client may require reducing network traffic and API calls.
+
+This paged result MUST return the following elements:
+
+- `digest`: The digest used to retrieve the referenced manifest
+- `manifest`: The [pretty](https://linuxhint.com/pretty_json_php/) listing of `oci.artifact.manifest`. By providing the manifest, consumers can choose which elements they require to determine which manifest of the paged result they must fetch. As the manifest is pretty formatted, the contents MAY not match the digest of the original manifest. For to assure the contents are accurate, the client MAY retrieve the manifest using `references.[n].digest`
+- `@nextLink`: Used for paged results
+
+As an example, Notary v2 manifests use annotations to determine which Notary v2 signature they should retrieve: `"org.cncf.notary.v2.signature.subject": "wabbit-networks.io"`
+
+**example result of artifacts that reference the `net-monitor` image:**
+```json
+{
+  "references": [
```

Comment:
Would this be cleaner as a list of descriptors? Returning the full manifest with (maybe) different formatting seems like it may require clients to make the second call to get the correctly formatted content anyways. Given that we're working with content addressable storage, I think I'd prefer to have registries return either descriptors / references to content or the exact content, but not this middle ground.

Reply:
I agree, this should just be a list of descriptors.

@SteveLasker you might be interested in a PR I put together to that would enable embedding the manifest in a single response: https://github.com/opencontainers/image-spec/pull/826

For an API like this, clients should be expected to fetch the original content, but registries would be able to embed `data`, if they want, so clients can avoid the extra roundtrip.

Reply:
From https://github.com/opencontainers/artifacts/pull/29#discussion_r632859455 it seems like they want to make this pretty for human consumption? Especially for an API, that doesn't feel like it's worth the added complexity of departing from OCI norms.

Reply:
We did initially use a list of descriptors, and we've iterated a few times on the results. Prototype-1 returned a list of descriptors, which required the client to pull each manifest as the digest alone doesn't provide enough info to know which artifact the client actually cares about. Consider there may be several referrers to an image: (original signature, sbom, scan-result (s), gpl source, additional signatures for content promotion)

Please see: [Artifact Referrers API results](https://github.com/opencontainers/artifacts/blob/39440f5f5c97f4a32ad0b4eba2357162161ab063/manifest-referrers-api.md#artifact-referrers-api-results)

> To avoid an artifact client from having to retrieve each manifest, just to determine if it's the specific artifact needed to continue processing, the /referrers API will return a collection of manifests, including the annotations within each manifest. By providing manifests, as opposed to manifest descriptors, a specific artifact client can find the relevant properties they need to determine which artifact to retrieve. For example, Notary v2 MAY use an annotation: "org.cncf.notary.v2.signature.subject": "wabbit-networks.io", which the client could use to determine which signature to pull from the registry. Using annotations, clients can reduce round trips and the data returned to determine which artifacts the specific client may require reducing network traffic and API calls.

Reply:
`referrers`?